File: packages/server-python/app/routers/games.py
"""Game and room routes."""

from datetime import datetime
from typing import Optional
from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import func
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import get_current_user
from app.models.game import Game, GamePlayer, GameRoom, GameState
from app.models.user import User
from app.schemas.game import (
    GameResponse,
    PlayerResponse,
    RoomCreate,
    RoomListResponse,
    RoomResponse,
)

router = APIRouter(prefix="/games", tags=["games"])


def _game_to_response(game: Game) -> GameResponse:
    """Convert Game model to GameResponse."""
    return GameResponse(
        id=game.id,
        name=game.name,
        description=game.description,
        minPlayers=game.min_players,
        maxPlayers=game.max_players,
        imageUrl=game.image_url,
        createdAt=game.created_at,
    )


def _player_to_response(player: GamePlayer) -> PlayerResponse:
    """Convert GamePlayer model to PlayerResponse."""
    return PlayerResponse(
        id=player.id,
        userId=player.user_id,
        position=player.position,
        score=player.score,
    )


@router.get("/", response_model=list[GameResponse])
def list_games(db: Session = Depends(get_db)) -> list[GameResponse]:
    """List all available games."""
    games = db.query(Game).all()
    return [_game_to_response(g) for g in games]


@router.get("/rooms", response_model=list[RoomListResponse])
def list_rooms(
    game_id: Optional[str] = Query(None, alias="gameId"),
    status_filter: Optional[str] = Query(None, alias="status"),
    db: Session = Depends(get_db),
) -> list[RoomListResponse]:
    """List rooms with optional filters."""
    query = (
        db.query(
            GameRoom.id,
            GameRoom.name,
            GameRoom.game_id,
            Game.name.label("game_name"),
            Game.image_url.label("game_image"),
            GameRoom.host_id,
            GameRoom.status,
            GameRoom.max_players,
            GameRoom.created_at,
            func.count(GamePlayer.id).label("player_count"),
        )
        .outerjoin(Game, GameRoom.game_id == Game.id)
        .outerjoin(GamePlayer, GameRoom.id == GamePlayer.room_id)
        .group_by(GameRoom.id)
    )

    if game_id:
        query = query.filter(GameRoom.game_id == game_id)

    if status_filter:
        query = query.filter(GameRoom.status == status_filter)

    results = query.all()

    return [
        RoomListResponse(
            id=r.id,
            name=r.name,
            gameId=r.game_id,
            gameName=r.game_name,
            gameImage=r.game_image,
            hostId=r.host_id,
            status=r.status,
            maxPlayers=r.max_players,
            playerCount=r.player_count,
            createdAt=r.created_at,
        )
        for r in results
    ]


@router.get("/rooms/{room_id}", response_model=RoomResponse)
def get_room(room_id: str, db: Session = Depends(get_db)) -> RoomResponse:
    """Get room details with players."""
    room = db.query(GameRoom).filter(GameRoom.id == room_id).first()
    if not room:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Room not found",
        )

    players = (
        db.query(GamePlayer)
        .filter(GamePlayer.room_id == room_id)
        .order_by(GamePlayer.position)
        .all()
    )

    return RoomResponse(
        id=room.id,
        name=room.name,
        gameId=room.game_id,
        hostId=room.host_id,
        status=room.status,
        maxPlayers=room.max_players,
        createdAt=room.created_at,
        startedAt=room.started_at,
        finishedAt=room.finished_at,
        players=[_player_to_response(p) for p in players],
    )


@router.post("/rooms", response_model=dict)
def create_room(
    room_data: RoomCreate,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Create a new room."""
    # Verify game exists
    game = db.query(Game).filter(Game.id == room_data.gameId).first()
    if not game:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Game not found",
        )

    room_id = str(uuid4()).replace("-", "")

    # Create room
    room = GameRoom(
        id=room_id,
        game_id=room_data.gameId,
        name=room_data.name,
        host_id=user.id,
        status="waiting",
        max_players=room_data.maxPlayers,
    )
    db.add(room)

    # Add host as first player
    player = GamePlayer(
        id=str(uuid4()).replace("-", ""),
        room_id=room_id,
        user_id=user.id,
        position=0,
    )
    db.add(player)

    db.commit()
    return {"roomId": room_id}


@router.post("/rooms/{room_id}/join", response_model=dict)
def join_room(
    room_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Join a room."""
    room = db.query(GameRoom).filter(GameRoom.id == room_id).first()
    if not room:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Room not found",
        )

    if room.status != "waiting":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Game already started",
        )

    # Check if already in room
    existing = (
        db.query(GamePlayer)
        .filter(GamePlayer.room_id == room_id, GamePlayer.user_id == user.id)
        .first()
    )
    if existing:
        return {"success": True, "message": "Already in room"}

    # Check room capacity
    player_count = db.query(GamePlayer).filter(GamePlayer.room_id == room_id).count()
    if player_count >= room.max_players:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Room is full",
        )

    # Add player
    player = GamePlayer(
        id=str(uuid4()).replace("-", ""),
        room_id=room_id,
        user_id=user.id,
        position=player_count,
    )
    db.add(player)
    db.commit()

    return {"success": True}


@router.post("/rooms/{room_id}/leave", response_model=dict)
def leave_room(
    room_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Leave a room."""
    db.query(GamePlayer).filter(
        GamePlayer.room_id == room_id, GamePlayer.user_id == user.id
    ).delete()
    db.commit()
    return {"success": True}


@router.post("/rooms/{room_id}/start", response_model=dict)
def start_game(
    room_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    """Start a game (host only)."""
    room = db.query(GameRoom).filter(GameRoom.id == room_id).first()
    if not room:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Room not found",
        )

    if room.host_id != user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only host can start the game",
        )

    if room.status != "waiting":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Game already started",
        )

    # Check player count
    player_count = db.query(GamePlayer).filter(GamePlayer.room_id == room_id).count()
    game = db.query(Game).filter(Game.id == room.game_id).first()

    if player_count < game.min_players:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Need at least {game.min_players} players to start",
        )

    # Update room status
    room.status = "playing"
    room.started_at = datetime.utcnow()

    # Create initial empty game state
    state = GameState(
        room_id=room_id,
        current_turn=0,
    )
    state.state = {"cards": [], "flipped": [], "matched": []}
    db.add(state)

    db.commit()
    return {"success": True}


@router.get("/{game_id}", response_model=GameResponse)
def get_game(game_id: str, db: Session = Depends(get_db)) -> GameResponse:
    """Get a specific game by ID."""
    game = db.query(Game).filter(Game.id == game_id).first()
    if not game:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Game not found",
        )
    return _game_to_response(game)

File: packages/platform/server/routes/games.ts
import { Elysia, t } from 'elysia'
import { db } from '../db/index.js'
import { games, gameRooms, gamePlayers, gameState } from '../db/schema.js'
import { eq, and, sql } from 'drizzle-orm'
import { randomBytes } from 'crypto'
import { authMiddleware } from '../middleware/auth.js'

function generateId() {
  return randomBytes(16).toString('hex')
}

export const gamesRoutes = new Elysia({ prefix: '/api/games' })
  .use(authMiddleware)
  .get('/', async () => {
    return await db.select().from(games)
  })
  .get(
    '/rooms',
    async ({ query }) => {
      let queryBuilder = db
        .select({
          id: gameRooms.id,
          name: gameRooms.name,
          gameId: gameRooms.gameId,
          gameName: games.name,
          gameImage: games.imageUrl,
          hostId: gameRooms.hostId,
          status: gameRooms.status,
          maxPlayers: gameRooms.maxPlayers,
          playerCount: sql<number>`count(${gamePlayers.id})`,
          createdAt: gameRooms.createdAt,
        })
        .from(gameRooms)
        .leftJoin(games, eq(gameRooms.gameId, games.id))
        .leftJoin(gamePlayers, eq(gameRooms.id, gamePlayers.roomId))
        .groupBy(gameRooms.id)
        .$dynamic()

      if (query.gameId) {
        queryBuilder = queryBuilder.where(eq(gameRooms.gameId, query.gameId))
      }

      if (query.status) {
        queryBuilder = queryBuilder.where(eq(gameRooms.status, query.status))
      }

      return await queryBuilder
    },
    {
      query: t.Object({
        gameId: t.Optional(t.String()),
        status: t.Optional(t.Union([t.Literal('waiting'), t.Literal('playing'), t.Literal('finished')])),
      }),
    },
  )
  .get(
    '/rooms/:roomId',
    async ({ params, set }) => {
      const room = await db
        .select()
        .from(gameRooms)
        .where(eq(gameRooms.id, params.roomId))
        .limit(1)

      if (!room[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Room not found' }
      }

      const players = await db
        .select({
          id: gamePlayers.id,
          userId: gamePlayers.userId,
          position: gamePlayers.position,
          score: gamePlayers.score,
        })
        .from(gamePlayers)
        .where(eq(gamePlayers.roomId, params.roomId))
        .orderBy(gamePlayers.position)

      return {
        ...room[0],
        players,
      }
    },
    {
      params: t.Object({
        roomId: t.String(),
      }),
    },
  )
  .post(
    '/rooms',
    async ({ body, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      const roomId = generateId()

      await db.insert(gameRooms).values({
        id: roomId,
        gameId: body.gameId,
        name: body.name,
        hostId: user.id,
        status: 'waiting',
        maxPlayers: body.maxPlayers,
      })

      await db.insert(gamePlayers).values({
        id: generateId(),
        roomId,
        userId: user.id,
        position: 0,
      })

      return { roomId }
    },
    {
      body: t.Object({
        gameId: t.String(),
        name: t.String({ minLength: 1, maxLength: 50 }),
        maxPlayers: t.Number({ minimum: 2, maximum: 6 }),
      }),
    },
  )
  .post(
    '/rooms/:roomId/join',
    async ({ params, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      const room = await db
        .select()
        .from(gameRooms)
        .where(eq(gameRooms.id, params.roomId))
        .limit(1)

      if (!room[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Room not found' }
      }

      if (room[0].status !== 'waiting') {
        set.status = 400
        return { error: 'Bad request', message: 'Game already started' }
      }

      const existingPlayer = await db
        .select()
        .from(gamePlayers)
        .where(and(eq(gamePlayers.roomId, params.roomId), eq(gamePlayers.userId, user.id)))
        .limit(1)

      if (existingPlayer[0]) {
        return { success: true, message: 'Already in room' }
      }

      const playerCount = await db
        .select({ count: sql<number>`count(*)` })
        .from(gamePlayers)
        .where(eq(gamePlayers.roomId, params.roomId))

      if (playerCount[0].count >= room[0].maxPlayers) {
        set.status = 400
        return { error: 'Bad request', message: 'Room is full' }
      }

      await db.insert(gamePlayers).values({
        id: generateId(),
        roomId: params.roomId,
        userId: user.id,
        position: playerCount[0].count,
      })

      return { success: true }
    },
    {
      params: t.Object({
        roomId: t.String(),
      }),
    },
  )
  .post(
    '/rooms/:roomId/leave',
    async ({ params, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      await db
        .delete(gamePlayers)
        .where(and(eq(gamePlayers.roomId, params.roomId), eq(gamePlayers.userId, user.id)))

      return { success: true }
    },
    {
      params: t.Object({
        roomId: t.String(),
      }),
    },
  )
  .post(
    '/rooms/:roomId/start',
    async ({ params, user, set }) => {
      if (!user) {
        set.status = 401
        return { error: 'Unauthorized', message: 'You must be logged in' }
      }

      const room = await db
        .select()
        .from(gameRooms)
        .where(eq(gameRooms.id, params.roomId))
        .limit(1)

      if (!room[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Room not found' }
      }

      if (room[0].hostId !== user.id) {
        set.status = 403
        return { error: 'Forbidden', message: 'Only host can start the game' }
      }

      if (room[0].status !== 'waiting') {
        set.status = 400
        return { error: 'Bad request', message: 'Game already started' }
      }

      const playerCount = await db
        .select({ count: sql<number>`count(*)` })
        .from(gamePlayers)
        .where(eq(gamePlayers.roomId, params.roomId))

      const game = await db.select().from(games).where(eq(games.id, room[0].gameId)).limit(1)

      if (playerCount[0].count < game[0].minPlayers) {
        set.status = 400
        return { error: 'Bad request', message: `Need at least ${game[0].minPlayers} players to start` }
      }

      await db
        .update(gameRooms)
        .set({
          status: 'playing',
          startedAt: new Date(),
        })
        .where(eq(gameRooms.id, params.roomId))

      const initialState = {
        cards: [],
        flipped: [],
        matched: [],
      }

      await db.insert(gameState).values({
        roomId: params.roomId,
        state: initialState,
        currentTurn: 0,
      })

      return { success: true }
    },
    {
      params: t.Object({
        roomId: t.String(),
      }),
    },
  )
  .get(
    '/:gameId',
    async ({ params, set }) => {
      const game = await db.select().from(games).where(eq(games.id, params.gameId)).limit(1)

      if (!game[0]) {
        set.status = 404
        return { error: 'Not found', message: 'Game not found' }
      }

      return game[0]
    },
    {
      params: t.Object({
        gameId: t.String(),
      }),
    },
  )
